### PR TITLE
Cost accounting I

### PR DESCRIPTION
**Pull request by marcin-rzeznicki**
## Overview
<sup>_What this PR does, and why it's needed_</sup>
Cost accounting
The groundwork for the cost accounting has been laid out by @ArturGajowy  in [this gist](https://gist.github.com/ArturGajowy/73e481767548b084a8e5d7c19bb1f70c/4766ea917285cb6909231ec25977993ad6d1ab1a). This PR is joint work between @JosDenmark and me

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>
https://rchain.atlassian.net/browse/RCHAIN-3767
https://rchain.atlassian.net/browse/RCHAIN-3624
https://rchain.atlassian.net/browse/RCHAIN-3625

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

Currently it contains the following:
- adds notion of SystemDeploy
- implements PreCharge
- implements Refund
- implements Slash
- implements CheckBalance
- adds type-safe inputs (NormalizerEnv)
- integrates with existing RuntimeManager flow to _play_ and _replay_ system deploys
- contains tests for basic scenarios (they're not completed yet but @JosDenmark has made sure that basic scenarios work)

In subsequent PRs we'll be adding:
- CloseBlock
- more tests
- charging logic

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [ ] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/test/scala/coop/rchain/casper/util/rholang/Interactive.scala [hunk 1]
```diff
@@ -54,7 +50,7 @@ class Interactive private (runtime: Runtime[Task])(implicit scheduler: Scheduler
   def checkpointNames: List[String] = checkpoints.keys.toList
 
   def eval(code: String): Unit = {
-    TestUtil.eval(code, runtime, NormalizerEnv.Empty).runSyncUnsafe(Duration.Inf)
+    TestUtil.eval(code, runtime, Map.empty).runSyncUnsafe(Duration.Inf)
```
Comment (tymm) [inline casper/src/test/scala/coop/rchain/casper/util/rholang/Interactive.scala:53]: Why do you prefer `Map.empty` over `NormalizerEnv.Empty`?

File: models/src/main/scala/coop/rchain/models/HashM.scala [hunk 2]
```diff
@@ -88,14 +86,15 @@ object HashM extends HashMDerivation {
...
 
-  implicit val ParHash: HashM[Par] = gen[Par]
-  implicit val ExprHash            = gen[Expr]
-  implicit val VarHash             = gen[Var]
-  implicit val SendHash            = gen[Send]
-  implicit val ReceiveHash         = gen[Receive]
-  implicit val ReceiveBindHash     = gen[ReceiveBind]
-  implicit val NewHash             = gen[New]
-  implicit val MatchHash           = gen[Match]
+  implicit val Env: HashM[Map[String, Par]] = opaqueHash
```
Comment (tymm) [inline models/src/main/scala/coop/rchain/models/HashM.scala:89]: Please see this discussion: https://github.com/rchain/rchain/pull/2763#discussion_r335401269
If possible we should stick to `HashM[NormalizerEnv]`.

File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/RevGenerator.scala [hunk 3]
```diff
@@ -1,63 +1,58 @@
...
-import coop.rchain.models.{NormalizerEnv, Par}
+import coop.rchain.models.NormalizerEnv
 import coop.rchain.rholang.build.CompiledRholangSource
-import coop.rchain.rholang.interpreter.ParBuilder
-import monix.eval.Coeval
-
-final case class RevGenerator(userVaults: Seq[Vault], supply: Long) extends CompiledRholangSource {
 
+final class RevGenerator private (supply: Long, code: String)
+    extends CompiledRholangSource(code, NormalizerEnv.Empty) {
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/genesis/contracts/RevGenerator.scala:7]: This and the changes related to this should be in a separate commit.

Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/genesis/contracts/RevGenerator.scala:7]: Definitely :+1:

File: casper/src/test/scala/coop/rchain/casper/util/rholang/Interactive.scala (hunk 1 above)
Comment (marcin-rzeznicki) [reply casper/src/test/scala/coop/rchain/casper/util/rholang/Interactive.scala:53]: I'm not sure if my reasoning holds water, please evaluate: Since `NormallizerEnv` is now type-safe - meaning it tracks the environment it contains, and, some pieces of code can't make use of it - I decided to split it. When you have to work with "just a map" you don't need to depend on `NormalizerEnv`. Only when you can take advantage of a precise `Env` type(like SD do), you should depend on `NornalizerEnv`. OTOH, it is possible to pass `NornalizerEnv.Empty` etc - but then you just type more (type var declaration, additional conversion to `.toEnv`) for no gain

File: models/src/main/scala/coop/rchain/models/HashM.scala (hunk 2 above)
Comment (marcin-rzeznicki) [reply models/src/main/scala/coop/rchain/models/HashM.scala:89]: I read that - but I don't understand still. Please explain - `NornalizerEnv` was just a type alias before

Comment (tymm) [reply models/src/main/scala/coop/rchain/models/HashM.scala:89]: It's just a little bit safer in terms of stack-safety because if someone one day creates a `map: Map[String, Par]` with some really big `Par` inside, Scala will pick up the `opaqueHash` instance and will blow up the stack.
With `HashM[NormalizerEnv]` that someone won't be able to do that. At least not without writing the non stack-safe `HashM` instance himself.

File: casper/src/test/scala/coop/rchain/casper/util/rholang/Interactive.scala (hunk 1 above)
Comment (tymm) [reply casper/src/test/scala/coop/rchain/casper/util/rholang/Interactive.scala:53]: I guess that makes sense. Thank you for the explanation.

File: models/src/main/scala/coop/rchain/models/HashM.scala (hunk 2 above)
Comment (marcin-rzeznicki) [reply models/src/main/scala/coop/rchain/models/HashM.scala:89]: Agreed

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala [hunk 4]
```diff
@@ -73,6 +77,70 @@ class RuntimeManagerImpl[F[_]: Concurrent: Metrics: Span: Log](
...
+  )(
+      implicit provided: NormalizerEnv.Provides[Env0, systemDeploy.Env],
+      ev: ToEnvMap[systemDeploy.Env]
+  ): F[Either[SystemDeployFailure, SystemDeployResult[systemDeploy.Result]]] = {
+    implicit val c: _cost[F] = runtime.cost
+    for {
+      softCheckpoint <- runtime.space.createSoftCheckpoint()
+      result <- EitherT
+                 .liftF(
+                   Interpreter[F]
```
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: This needs to be a `Interpreter[F].injAttempt` call, and we should pull `c: _cost[F]` into the `liftF` scope. That's also where we'll have to provide the `rand` implicit.

Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: `into the liftF scope` ?

Comment (JosephDenman) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: The minimal scope. 

```
EitherT.liftF {
   implicit c: _cost[F] = runtime.cost
   Interpreter[F].injAttempt(...)
}
```

Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: Fixed in 7ca418a

File: casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala [hunk 5]
```diff
@@ -23,6 +26,11 @@ abstract class SystemDeploy(val rand: Blake2b512Random) {
   protected def toEnvMap: ToEnvMap[Env]
   final def env: Map[String, Par] = normalizerEnv.toEnv(toEnvMap)
 
+  implicit protected val envsReturnChannel: Contains.Aux[Env, `sys:casper:return`, GUnforgeable]
+  protected def mkReturnChannel =
```
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala:31]: It's interesting that it is critical that this is a definition.

Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala:31]: You might get NPE otherwise (though I didn't check). When a class is initialized on JVm it's superclasses are initialized first. Had this been `val`, `envsReturnChannel` would still be `null`. Boom.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala [hunk 6]
```diff
@@ -31,14 +29,12 @@ final class PreChargeDeploy(chargeAmount: Long, pk: PublicKey, rand: Blake2b512R
       (`sys:casper:return` ->> GUnforgeable) :: HNil
 
   import toPar._
-  protected override val toEnvMap = ToEnvMap[Env]
+  protected override val envsReturnChannel = Contains[Env, `sys:casper:return`]
```
Comment (JosephDenman) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala:28]: Can we de-duplicate `envsReturnChannel` from `PreChargeDeploy` and `RefundDeploy` by defining it in `SystemDeploy` using the `Env` existential?

Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala:28]: Unfortunately, I don't think it would work. In `SystemDeploy` `Env` is, as you noted, existential which means that it's structure is not known. When you take a look at the implicit expansion at, say, `PreChargeDeploy` you will notice that the expression `Contains[Env, returnChannel]` expands to `NormalizerEnv.Contains.Aux[(sys:casper:deployerId ->> GDeployerId) :: (sys:casper:invalidBlockHash ->> Expr.ExprInstance.GString) :: (sys:casper:return ->> GUnforgeable) :: HNil,sys:casper:return,ev.KeyType]` - this is precisely the info we're missing in `SystemDeploy` . I don't see how it would work :disappointed:

Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala:28]: PS. I guess my point is that even though the expression looks the same, it actually expands to different things (using knowledge present at the point of expansion).

File: casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  import NormalizerEnv._
+
+  type Output
+  type Result
+  type Env
+
+  final val `sys:casper:initialDeployerId` = Witness("sys:casper:initialDeployerId")
+  final val `sys:casper:return`            = Witness("sys:casper:return")
+
+  type `sys:casper:initialDeployerId` = `sys:casper:initialDeployerId`.T
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala]: What exactly is this for? I assume it's the user's deployer id? I don't think every SD will need this.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  type `sys:casper:return`            = `sys:casper:return`.T
+
+  def source: String
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala]: I thought the idea was to use a `templateId` and not have code in the SDs?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/package.scala [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package coop.rchain.casper.util.rholang
+
+import shapeless.labelled.FieldType
+
+package object costacc {
+  type ->>[A, B] = FieldType[A, B]
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/package.scala:6]: This is something different than the `->>` from `shapeless.syntax.singleton`?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  import syntax.singleton._
+  import record._
+
+  type Output = (RhoBoolean, RhoString)
+  type Result = Unit
+
+  val `sys:casper:chargeAmount` = Witness("sys:casper:chargeAmount")
+  type `sys:casper:chargeAmount` = `sys:casper:chargeAmount`.T
+
+  type Env =
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala:24]: Seeing it in action I'm no longer sure that all this additional shapeless magic is worth the benefit 😬

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        ("sys:casper:chargeAmount" ->> GInt(chargeAmount)) ::
+        ("sys:casper:return" ->> GUnforgeable(
+          GPrivateBody(GPrivate(ByteString.copyFrom(rand.next())))
+        )) :: HNil
+    )
+
+  override val source: String =
+    """
+                                |new rl(`rho:registry:lookup`),
+                                |poSCh,
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala]: Please align like that:
```
new rl(`rho:registry:lookup`),
    poSCh,
    initialDeployerId
    ...
in {
```
and use `#` instead of `|`.

**Review by tymm — COMMENTED** [review]
First couple of comments. More will follow.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala (hunk 7 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala]: Hmm, this is outdated. I am not sure why this thing shows up. "Fixed" in 4cb5d212502306738871a89fce590dec3c913e17

File: casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala (hunk 8 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala]: Well, the `templateId` is the type (of SystemDeploy) and `code` is a "mapping" between SD and corresponding code. Is this interpretation OK?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/package.scala (hunk 9 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/package.scala:6]: I promoted shapeless' `->>` to a type for easier writing of `Env` types. So, just as value level `->>` creates a `FieldType` with appropriate singleton type - this type constructor create a type that value level `->>` would create. It is used just for easy writing of SD specs

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala (hunk 10 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala:24]: Yeah, maybe - but:
1. we castrated this part a bit (we had more fun stuff like checking at compile type if you can combine environments etc)
1. CA correctness is important so it's nice to know that your "strings" match
1. I'd love to expand it virally through the system - I was actually astonished that string-based programming prevails here :-)
1. we'll have more use cases later (eg. asserting at compile time you're using only `sys` namespace etc)
1. people who'll come after us will either cry and make their SDs literally verified by the compiler or tear it down - but no sloppiness will creep out here :-)
1. In Scala 2.13 this is first-class so you can probably throw away shapeless

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala [hunk 12]
```diff
@@ -35,27 +34,25 @@ object PreChargeDeploy extends SystemDeploy {
...
-                                |chargeAmount(`sys:casper:chargeAmount`),
-                                |return(`sys:casper:return`) in
-                                |{
-                                |  rl!(`rho:rchain:pos`, *poSCh) |
-                                |  for(@(_, PoS) <- poSCh) {
-                                |    @PoS!("chargeDeploy", *initialDeployerId, *chargeAmount, *return)
-                                |  }
-                                |
-                                |}""".stripMargin
+    """|new rl(`rho:registry:lookup`),
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala]: The same from my previous comment about alignment applies here as well.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala (hunk 8 above)
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeploy.scala]: Oh, makes sense. What we end up sending over the wire are `ProcessedSystemDeploy`, right? And they won't contain the `source`.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala [hunk 13]
```diff
@@ -74,6 +86,171 @@ class RuntimeManagerImpl[F[_]: Concurrent: Metrics: Span: Log](
...
+            consumeResult(runtime)(systemDeploy, replay = false) >>= {
+              case Some((_, Seq(ListParWithRandom(Seq(par), _)))) =>
+                systemDeploy.extractResult(par) match {
+                  case Right(result) =>
+                    runtime.space
+                      .createSoftCheckpoint()
+                      .map(
+                        postDeploySoftCheckpoint => (postDeploySoftCheckpoint.log, result.asRight)
+                      )
+                  case Left(systemDeployError: SystemDeployError) =>
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:237]: An example for this case would be a `chargeDeploy` SD that failed because of insufficient funds, correct?
Why can we revert to `preDeploySoftCheckpoint`? Couldn't this cause problems with replayability?

Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:237]: Yes  like `insufficient funds`. Hmm, we're facing some errors in replays. Maybe your comment is spot on. @JosDenmark could you please take a look?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala (hunk 11 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala]: Fixed in 31424de3

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala (hunk 12 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala]: Also, fixed in 31424de3

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/RefundDeploy.scala [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+  def createNormalizerEnv(refundAmount: Long, rand: Blake2b512Random) =
+    new NormalizerEnv(
+      ("sys:casper:refundAmount" ->> GInt(refundAmount)) ::
+        ("sys:casper:return" ->> GUnforgeable(
+          GPrivateBody(GPrivate(ByteString.copyFrom(rand.next())))
+        )) :: HNil
+    )
+
+  override val source: String =
+    """|new rl(`rho:registry:lookup`),
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/RefundDeploy.scala]: Please see previous comment about alignment.

File: models/src/main/scala/coop/rchain/models/rholang/implicits.scala [hunk 15]
```diff
@@ -284,6 +284,9 @@ object implicits {
         None
       }
 
+    def isNil(): Boolean =
+      p.sends.isEmpty && p.receives.isEmpty && p.news.isEmpty && p.matches.isEmpty && p.bundles.isEmpty && p.exprs.isEmpty
```
Comment (tymm) [inline models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: Shouldn't we add `&& p.unforgeables.isEmpty && p.connectives.isEmpty`?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala [hunk 16]
```diff
@@ -31,10 +31,17 @@ import coop.rchain.rholang.interpreter.{
...
-import coop.rchain.rspace.{Blake2b256Hash, ReplayException, Result}
+import coop.rchain.rspace.{Blake2b256Hash, ReplayException}
 import coop.rchain.shared.Log
 
 trait RuntimeManager[F[_]] {
+  def playSystemDeploy[S <: SystemDeploy](startHash: StateHash)(
+      systemDeploy: S
+  ): F[SystemDeployResult[systemDeploy.Result]]
+  def replaySystemDeploy[S <: SystemDeploy](startHash: StateHash)(
+      systemDeploy: S,
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:43]: Just want to be sure that I get it right: every node has to generate system deploys from user deploys and then checks these against `ProcessedSystemDeploy`s, correct?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    new NormalizerEnv(
+      ("sys:casper:deployerId" ->> GDeployerId(ByteString.copyFrom(pk.bytes))) :: ("sys:casper:invalidBlockHash" ->> GString(
+        invalidBlockHash.base16String
+      )) :: mkReturnChannel :: HNil
+    )
+
+  override val source: String =
+    """|new rl(`rho:registry:lookup`),
+       |poSCh,
+       |deployerId(`sys:casper:deployerId`),
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala]: Before I had seen `initialDeployerId`. Could we make it `deployerId` in all SDs?

Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala]: Or maybe here it would even make sense to name it `validatorDeployerId` for more clarity.

File: casper/src/test/scala/coop/rchain/casper/util/rholang/RuntimeManagerTest.scala [hunk 18]
```diff
@@ -201,11 +282,60 @@ class RuntimeManagerTest extends FlatSpec with Matchers {
     } yield ()
   }
 
+  "computeState" should "be replayed by replayComputeState" in effectTest {
+    runtimeManagerResource.use { runtimeManager =>
+      for {
+        deploy <- ConstructDeploy.sourceDeployNowF(
+                   """
+                                                            # new deployerId(`rho:rchain:deployerId`),
```
Comment (tymm) [inline casper/src/test/scala/coop/rchain/casper/util/rholang/RuntimeManagerTest.scala:290]: Alignment seems a bit off here.

**Review by tymm — APPROVED** [review]
Great work! Thanks!

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala (hunk 17 above)
Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala]: And in the others `userDeployerId` would maybe make sense. But up to you guys.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/RefundDeploy.scala (hunk 14 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/RefundDeploy.scala]: Yep, fixed in 31424de361c2e2ba1351182e102a76e2dbcac1b4

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala (hunk 16 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:43]: Correct!

File: casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala (hunk 17 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/SlashDeploy.scala]: Yeah, as I commented: you'd seen the outdated version where we'd had the names mixed up. Now it's `deployerId` everywhere. Makes sense?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala (hunk 13 above)
Comment (marcin-rzeznicki) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:237]: While waiting for your assessment I removed these doubtful reverts in 0b62efd50073afc1c958d6a5642b3042a7cc92be

File: models/src/main/scala/coop/rchain/models/rholang/implicits.scala (hunk 15 above)
Comment (marcin-rzeznicki) [reply models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: Ah, correct, thanks!

Comment (marcin-rzeznicki) [reply models/src/main/scala/coop/rchain/models/rholang/implicits.scala]: Fixed in b4829c4c8f6b0c8cc2a55d0a7dc2735aff2e97b1

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala [hunk 19]
```diff
@@ -74,6 +86,171 @@ class RuntimeManagerImpl[F[_]: Concurrent: Metrics: Span: Log](
...
+    runtime.replaySpace.rig(processedSystemDeploy.eventList.map(EventConverter.toRspaceEvent)) >> {
+      runtime.replaySpace.createSoftCheckpoint() >>= { softCheckpoint =>
+        evaluateSystemSource(runtime)(systemDeploy, replay = true) >>= {
+          case EvaluateResult(_, interpreterErrors) =>
+            if (interpreterErrors.nonEmpty) UnexpectedSystemErrors(interpreterErrors).raiseError
+            else
+              consumeResult(runtime)(systemDeploy, replay = true) >>= { consumeResultOpt =>
+                runtime.replaySpace.checkReplayData().attempt.map(_.swap.toOption) >>= {
+                  case Some(replayException: ReplayException) =>
+                    ReplayFailure.unusedCOMMEvent(replayException).asLeft.pure
```
Comment (zsluedem) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:173]: Do we need `revertToSoftCheckpoint ` here?

Comment (tymm) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:173]: I think we should add some comments to the code about when and why we revert or not. It's hard to understand from code.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala (hunk 13 above)
Comment (JosephDenman) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:237]: Timm is right. In a forthcoming PR we fix these issues, reverting only when an user error (`InterpreterError`) is thrown during `replayDeploy` or `processDeploy`.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala (hunk 19 above)
Comment (JosephDenman) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala:173]: @zsluedem https://github.com/rchain/rchain/pull/2778#discussion_r341574398

**Review by JosephDenman — APPROVED** [review]
